hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/NodeStatusUpdaterImpl.java

```diff
@@ -717,6 +717,8 @@ public void removeOrTrackCompletedContainersFromContext(
         org.apache.hadoop.yarn.server.nodemanager.containermanager.container.ContainerState.DONE)) {
         context.getContainers().remove(containerId);
         removedContainers.add(containerId);
+        // For the finished containers, the residual data stored in leveldb should also be cleared
+        addCompletedContainer(containerId);
```

Comment:
From a personal point of view, I think we should not clean up during the heartbeat. If the leveldb is not cleaned up successfully or an exception is thrown, is the heartbeat successful or failed?

The heartbeat is mainly to report container information and trigger scheduling, it is best not to clean up.

Reply:
Thanks for your review.
Cleaning up here is just be added into a map which will be handled asynchronously and no exception could be thrown here.

Reply:
Thanks for the explanation, but it's not needed for this operation, it doesn't seem like what heartbeat is supposed to do, `NodeStatusUpdaterImpl#addCompletedContainer` already does it, I think it's enough

Reply:
> Thanks for the explanation, but it's not needed for this operation, it doesn't seem like what heartbeat is supposed to do, `NodeStatusUpdaterImpl#addCompletedContainer` already does it, I think it's enough

When restarting the NM, we encountered an OOM issue during the recovery process. Upon analyzing the cause of the OOM in the heap memory, we found a large number of completed containers, approximately 150,000+. 

We suspect this is due to improper deletion of data in levelDB. Therefore, we believe that in addition to the memory deletion, the heartbeat response should also include a deletion operation for levelDB. 

Currently, the heartbeat response only performs memory deletion and does not handle levelDB deletion, which may lead to inconsistencies between the two deletion processes. Furthermore, the levelDB deletion in the heartbeat response is asynchronous, where the delete tasks are simply added to a pending deletion cache, so it does not affect the success of the heartbeat response.
![image](https://github.com/user-attachments/assets/e4e64aa7-e6cf-49b4-9e20-c6a387ce0a1c)